### PR TITLE
commands: project: Only update west/manifest via fast-forward

### LINKED ISSUE #59: West update should be skipped if west or manifest is on a feature branch
The current auto-update behavior could use a bit of tweaking to make it more intuitive to those users who are editing their manifest or west repository and don't want to type --no-update all the time.

The issue is that it's not very convenient to switch to a manifest branch with custom changes and then issue 'west fetch', 'west pull', etc. commands, as that would try to update the manifest itself as well, which isn't what the user wants IMO. We could emulate the repo behavior instead, which is only to update the manifest itself to incorporate remote changes if the manifest repository is on the default branch tracking the remote which is created by west init.

This applies at least to the manifest, though for consistency it could be applied to west as well.

### LINKED ISSUE #62: West update can leave manifest and west in unusable states
As a user, I expect that if `west update` fails (either run directly or as implied by `west fetch` or `pull`), that my installation is still in a usable state. However, the current implementation can leave either of these repositories in a broken (half-rebased) state if the update fails, with git merge conflict markers in source files.

To reproduce, just commit any changes that won't rebase cleanly and run west update.

The behavior instead should be to see if the rebase is possible to complete cleanly *without* damaging the existing installation, reporting an error if it fails with additional information on how to proceed.

### PR DESCRIPTION
**Pull request by ulfalizer**
If either the west or the manifest repository has local modifications,
skip automatically updating the modified repository, with a warning.
This avoids merge conflicts when working on them.

Maybe the manifest repository update could be skipped if the west
repository is out of date, and vice versa, but I suspect this version
might be good enough.

@ulfalizer please check if the issues below are addressed indeed.

Fixes #62 
Fixes #59

### CODE REVIEW

**Comment by tejlmand** [comment]
> Maybe the manifest repository update could be skipped as well if the
> manifest repository is out of date, and vice versa, but I suspect this
> version might be good enough.

You probably mean: _Maybe the manifest repository update could be skipped as well if the_ **west** _repository is out of date_, in the sentence above ?

besides that, I believe this PR to be good first step.
And then, let's take user requests from here, in case they don't like e.g. `west` to be update if `manifest` update fails.

**Comment by ulfalizer** [comment]
> You probably mean: _Maybe the manifest repository update could be skipped as well if the_ **west** _repository is out of date_, in the sentence above ?

Ah, yeah, typo. :)

File: src/west/commands/project.py [hunk 1]
```diff
@@ -687,19 +692,27 @@ def _update_manifest():
...
-        # Fetch changes from upstream
-        _git(project, 'fetch --quiet (remote) -- (revision)')
+        old_sha = _sha(project, 'HEAD')
+
+        # Only update special repositories if possible via fast-forward, as
+        # automatic rebasing is probably more annoying than useful when working
+        # directly on them. --rebase=false must be passed for --ff-only to be
+        # respected e.g. when pull.rebase is set.
+        if _git(project,
+                'pull --quiet --rebase=false --ff-only -- (remote) (revision)',
```
Comment (mbolivar) [inline src/west/commands/project.py]: I've tested that this works whether HEAD is detached or pointing at a branch.

File: src/west/commands/project.py [hunk 2]
```diff
@@ -687,19 +692,27 @@ def _update_manifest():
 
 
 def _update_special(name):
-    project = _special_project(name)
-
     with _error_context(_FAILED_UPDATE_MSG):
+        project = _special_project(name)
```
[the remark this answers is not in the document]
Comment (ulfalizer) [reply src/west/commands/project.py]: Unrelated refactoring. It makes more sense there, since it's related to updating special projects.

**Pull request merged**